FtpProvider/FtpProvider/FtpProvider.fs

```diff
@@ -1,102 +1,161 @@
+                                                                // TODO.3
+
+                                                              @@>))
+                        nestedType.AddMember contentsProperty
+
+                        yield nestedType :> MemberInfo 
+                   ]
+                   )
+        let actualType = ProvidedTypeDefinition(asm, nameSpace, typeName, Some typeof<obj>)
+        addTypes(site, actualType)
+        actualType
+
+    let addProvidedStaticParameter nme typ xmldoc = 
+      let p = ProvidedStaticParameter(nme,typ) 
+      p.AddXmlDoc(sprintf xmldoc)
+      p
+
+    let _ = 
+        let a = ProvidedTypeDefinition(asm, nameSpace, "FtpProvider", Some typeof<obj>)
+        a.AddXmlDoc("An FTP Type Provider which lets you 'dot' into directory structures, and then retrieve a file by 'dotting' into the '.Contents' tag.  note: there are no progress updates, so if it's a large file over a slow connection, the only solution is to wait.  Perhaps try a smaller file first to verify.")  // BUG.2 
+
+        let topType = ProvidedTypeDefinition(asm, nameSpace, "FtpProvider", Some typeof<obj>)
+        let siteParam = 
+           let p = ProvidedStaticParameter("Url",typeof<string>,"") 
+           p.AddXmlDoc(sprintf "The URL of the FTP site, including ftp://")
+           p
+        let userParam = 
+           let p = ProvidedStaticParameter("User",typeof<string>, "anonymous") 
+           p.AddXmlDoc("The user of the FTP site (default 'anonymous')")
+           p
+        let pwdParam = 
+           let p = ProvidedStaticParameter("Password",typeof<string>, "janedoe@contoso.com") 
+           p.AddXmlDoc("The password used to access the FTP site (default 'janedoe@contoso.com')")
+           p
+        let useBinary = 
+            let p = ProvidedStaticParameter("UseBinary",typeof<bool>, true)
+            p.AddXmlDoc("sets the data transfer data type to be binary (true) or ascii (false).  Binary mode gives a true, exact representation.  More often than not is the safer thing to use as this mode will handle both text and binary.  Use Ascii mode if you want to transfer text only, and you are happy to let FTP decide on appropriate line break characters translations, etc.")
+            p
+        let staticParams = [ siteParam; userParam; pwdParam; useBinary ]
+        topType.DefineStaticParameters(staticParams, (fun typeName args -> 
+            let site = args.[0] :?> string
+            let user = args.[1] :?> string
+            let pwd =  args.[2] :?> string
+            let useBinary = args.[3] :?> bool
+            createTypes(typeName, site, user, pwd, useBinary)))  // pass in top type details
+        this.AddNamespace(nameSpace, [topType])
+
+[<assembly:TypeProviderAssembly>]
+do ()
+
+// TODO
+// ----
+
+// TODO.1: pick up FTP courtesy details from environment variables?
+// TODO.2: add progress updates via intellisense?  ie. |          |  and |****  underneath for a text based left to right gauge

```

Comment:
Asynchronous loading like this could indeed be useful but probably more so at runtime (loading big files) rather than at compile-time (getting directory contents). Adding a ``GetContentsAsync()`` method (and changing ``Contents`` property to a ``GetContents()`` method for consistency) may be the right way to do this.